src/chainstate/burn/db/burndb.rs

```diff
@@ -421,363 +439,150 @@ const BURNDB_SETUP : &'static [&'static str]= &[
-            first_block_height: first_block_height,
-            first_burn_header_hash: first_burn_hash.clone(),
-        };
-
-        Ok(db)
+impl SortitionContext for SortitionDBTxContext {
+    fn first_block_height(&self) -> u64 {
+        self.first_block_height
     }
+}
 
-    /// Open a burn database in memory (used for testing)
-    #[cfg(test)]
-    pub fn connect_test(first_block_height: u64, first_burn_hash: &BurnchainHeaderHash) -> Result<BurnDB, db_error> { 
-        let mut rng = rand::thread_rng();
-        let mut buf = [0u8; 32];
-        rng.fill_bytes(&mut buf);
-        let db_path_dir = format!("/tmp/test-blockstack-burndb-{}", to_hex(&buf));
-        let (db_path, index_path) = db_mkdirs(&db_path_dir)?;
-
-        debug!("Open {}", &db_path);
-        let mut conn = Connection::open_with_flags(&db_path, OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE).map_err(db_error::SqliteError)?;
-        conn.busy_handler(Some(tx_busy_handler)).map_err(db_error::SqliteError)?;
-
-        BurnDB::instantiate(&mut conn, &index_path, first_block_height, first_burn_hash, get_epoch_time_secs())?;
-
-        let marf = BurnDB::open_index(&index_path)?;
-
-        let db = BurnDB {
-            conn: conn,
-            marf: marf,
-            readwrite: true,
-            first_block_height: first_block_height,
-            first_burn_header_hash: first_burn_hash.clone(),
-        };
-        
-        Ok(db)
+fn get_ancestor_sort_id<C: SortitionContext>(ic: &IndexDBConn<'_, C, SortitionId>, block_height: u64, tip_block_hash: &SortitionId) -> Result<Option<SortitionId>, db_error> {
+    let first_block_height = ic.context.first_block_height();
+    if block_height < first_block_height {
+        return Ok(None);
     }
+ 
+    ic.get_ancestor_block_hash(block_height - first_block_height, &tip_block_hash)
+}
 
-    /// Open the database on disk.  It must already exist and be instantiated.
-    /// It's best not to call this if you are able to call connect().  If you must call this, do so
-    /// after you call connect() somewhere else, since connect() performs additional validations.
-    pub fn open(path: &str, readwrite: bool) -> Result<BurnDB, db_error> {
-        let open_flags =
-            if readwrite {
-                OpenFlags::SQLITE_OPEN_READ_WRITE
-            }
-            else {
-                OpenFlags::SQLITE_OPEN_READ_ONLY
-            };
-
-        let (db_path, index_path) = db_mkdirs(path)?;
-        debug!("Open burndb '{}' as '{}', with index as '{}'",
-               db_path, if readwrite { "readwrite" } else { "readonly" }, index_path);
-        
-        let conn = Connection::open_with_flags(&db_path, open_flags).map_err(db_error::SqliteError)?;
-        conn.busy_handler(Some(tx_busy_handler)).map_err(db_error::SqliteError)?;
-
-        let marf = BurnDB::open_index(&index_path)?;
-        let first_snapshot = BurnDB::get_first_block_snapshot(&conn)?;
+pub struct SortitionId(pub [u8; 32]);
```

Comment:
I'm trying to wrap my head around how the use-case for `SortitionId` differs from `BurnchainHeaderHash` and `ConsensusHash`.  Can you verify my understanding?

My current understanding is that `SortitionId` identifies the sequence of burnchain operations accepted within a PoX fork.  So, there can be multiple `SortitionId`s per burnchain block.  You'd use a given `PoxIdentifier` to determine which `SortitionId` to use to identify the burnchain operations that actually get used in a given burnchain block.  In other words, the `(PoxIdentifier, SortitionId)` uniquely identifies a particular sortition in a burnchain block's (hence using that to identify a `BlockSnapshot`).  Is that all correct?

Reply:
The reason I ask is because looking forward, I'm wondering of there should be a single type for representing `(PoxIdentifier, SortitionId)` (kind of like how we use a `StacksBlockId` to identify a specific `(BurnchainHeaderHash, BlockHeaderHash)`).  We're doing to need something like this at some level of abstraction, because when I go to refactor the Stacks block chain state, its headers MARF is going to need to turn `(PoxIdentifier, SortitionId)` into a `MarfTrieId` somehow.

Reply:
Right -- so, a SortitionId uniquely identifies a sortition (which is the set of burnchain operations that are valid in a given block). This is also uniquely identified by `(PoxId, BurnchainHeaderHash)`, the `SortitionId` is just a `MarfTrieId` representation of that pair (it's just hashing that pair if the PoxId is non-zero).

For producing a `StacksBlockId`, currently that's done using `BurnchainHeaderHash` and `BlockHeaderHash` (the Stacks block header hash), but that would just need to be updated to `(SortitionId, BlockHeaderHash)`.

Reply:
Added some description of this and `PoxId` to their docstrings in b8b7abbcd

Reply:
Ah, that corrects a misunderstanding of mine.  It wasn't clear to me that `SortitionId` is already derived from `(PoxId, BurnchainHeaderHash)`.